Bottom: d765f6dc0453a8f3e3e782dd46baf8d21ae603fc
Top:    85de5466117b8b59abb9c8922040683276d40c9c
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2016-02-24 20:19:07 +0000

Limit virtual memory size of PCB process to avoid swapping my machine to death!

Sometimes the polygon code exploded...


---

diff --git a/src/polygon.c b/src/polygon.c
index 4dd9dc8..e58f7db 100644
--- a/src/polygon.c
+++ b/src/polygon.c
@@ -104,6 +104,10 @@ dicer output is used for HIDs which cannot render things with holes
 #include <dmalloc.h>
 #endif
 
+/* For getrlimit, setrlimit */
+#include <sys/time.h>
+#include <sys/resource.h>
+
 #define ROUND(x) ((long)(((x) >= 0 ? (x) + 0.5  : (x) - 0.5)))
 
 #define UNSUBTRACT_BLOAT 10
@@ -116,6 +120,8 @@ static double bw_rotate_circle_seg[4];
 void
 polygon_init (void)
 {
+  struct rlimit limit;
+
   double cos_ang = cos (2.0 * M_PI / POLY_CIRC_SEGS_F);
   double sin_ang = sin (2.0 * M_PI / POLY_CIRC_SEGS_F);
 
@@ -124,6 +130,11 @@ polygon_init (void)
 
   bw_rotate_circle_seg[0] =  cos_ang;  bw_rotate_circle_seg[1] =  sin_ang;
   bw_rotate_circle_seg[2] = -sin_ang;  bw_rotate_circle_seg[3] =  cos_ang;
+
+  /* DEBUG - AVOID PCB running the system out of memory! */
+  getrlimit (RLIMIT_AS, &limit);
+  limit.rlim_cur = MIN (limit.rlim_cur, 2000 * 1024 * 1024 /* 2 GiB limit to virtual memory size */);
+  setrlimit (RLIMIT_AS, &limit);
 }
 
 Cardinal
